src/dotty/tools/dotc/transform/TreeExtractors.scala

```diff
@@ -0,0 +1,54 @@
+package dotty.tools.dotc
+package transform
+
+import ast.{Trees, tpd}
+import core._, core.Decorators._
+import Contexts._, Flags._, Trees._, Types._, StdNames._, Symbols._
+import ValueClasses._
+
+object TreeExtractors {
+  import tpd._
+
+  /** Match arg1.op(arg2) and extract (arg1, op.symbol, arg2) */
+  object BinaryOp {
+    def unapply(t: Tree)(implicit ctx: Context): Option[(Tree, Symbol, Tree)] = t match {
+      case Apply(sel @ Select(arg1, _), List(arg2)) =>
+        Some((arg1, sel.symbol, arg2))
+      case _ =>
+        None
+    }
+  }
+
+ /** Match new C(args) and extract (C, args) */
+  object NewWithArgs {
+    def unapply(t: Tree)(implicit ctx: Context): Option[(Type, List[Tree])] = t match {
+      case Apply(Select(New(_), nme.CONSTRUCTOR), args) =>
+        Some((t.tpe, args))
+      case _ =>
+        None
+    }
+  }
+
+  /** For a value class like:
+   *    class V(val underlying: X) extends AnyVal
+   *  Match v.underlying() and extract v
+   *
+   *  Additionally, for a value class which is also a case class like:
+   *    case class V(val underlying: X) extends AnyVal
+   *  Match v._1() and extract v
+   */
+  object ValueClassUnbox {
+    def unapply(t: Tree)(implicit ctx: Context): Option[Tree] = t match {
+      case Apply(sel @ Select(ref, _), Nil) =>
+        val d = ref.tpe.typeSymbol.denot
+        if (isDerivedValueClass(d)
+          && (sel.symbol eq valueClassUnbox(d.asClass))
+          || ((d is Case) && (sel.symbol.name eq nme._1))) {
```

Comment:
https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/backend/jvm/DottyBackendInterface.scala#L692-L718
They implement those traits that make them use name-based pattern matching
https://github.com/DarkDimius/scala/blob/sharing-backend/src/compiler/scala/tools/nsc/backend/jvm/BackendInterface.scala#L225-L252 

As you can see, theses classes all have only a single field. If they were not implementing interfaces, I would want them to be value classes. As you can see, `_1` here has nontrivial implementation.